File: daemon/src/statemachined/daemon/mdns_service_advertisement.py
# SPDX-License-Identifier: LGPL-3.0-or-later
"""Telling the network this rig exists. docs/developer/daemon.md §5.

vstimd advertises `_vstimd._tcp` with a stable `id=` TXT record *"so clients can
match a device across name collisions"*. This is the same record for
`_statemachined._tcp`, and it exists for the same reason: a Pi whose hostname is
generated at boot cannot be hand-configured into a console's URL list, and two
rigs called `raspberrypi.local` is the normal case rather than the pathological
one.

**The identity is not the hostname.** `/etc/machine-id` survives a rename and a
DHCP lease and is different on every box, so a console that has seen this rig
before recognises it after somebody renames it. It is hashed rather than
published: a machine-id is meant to be treated as confidential, and what a
console needs is a *stable* identifier, not that particular one.

**Advertising is never fatal.** A daemon whose Avahi is not running, whose
network is down, or which cannot resolve an address of its own still owns a
device and still serves the API somebody may reach by IP. So every failure here
is logged and swallowed, and `is_advertising` says which happened rather than
leaving a caller to infer it.
"""

from __future__ import annotations

import hashlib
import socket
from dataclasses import dataclass, field
from pathlib import Path

#: The service type. `_statemachined._tcp` beside vstimd's `_vstimd._tcp` and
#: triald's, so one browse of the local domain finds every braemons daemon on a
#: rig.
SERVICE_TYPE = "_statemachined._tcp.local."

MACHINE_ID_PATH = Path("/etc/machine-id")


def stable_rig_identifier(machine_id_path: Path = MACHINE_ID_PATH) -> str:
    """Sixteen hex digits that mean "this box", across renames and reboots.

    Falls back to the hostname when there is no machine-id -- a container, a
    Mac, a developer's checkout. That is a weaker promise (a rename changes it),
    and it is the right fallback anyway: something stable-ish beats a value that
    changes every restart, which would make a console show one rig as many.
    """
    try:
        seed = machine_id_path.read_text().strip()
    except OSError:
        seed = ""
    if not seed:
        seed = socket.gethostname()
    return hashlib.sha256(f"statemachined:{seed}".encode()).hexdigest()[:16]


def text_records_for(
    *,
    rig_identifier: str,
    port: int,
    device_target: str,
    version: str,
) -> dict[bytes, bytes]:
    """What a console can act on without opening a connection first.

    Deliberately small. A TXT record is not an API: everything a console needs
    beyond "which rig, where, and can I speak its version" is one `GET
    /api/device` away, and duplicating that here would give a browser two
    answers that disagree the moment a board is unplugged.

    `path` is here because the elements contract (docs/developer/daemon.md §5) is a URL a
    console has to construct, and constructing it from a convention rather than
    from the record is how a console breaks when the daemon is behind a proxy.
    """
    return {
        b"id": rig_identifier.encode(),
        b"version": version.encode(),
        b"api": b"/api",
        b"elements": b"/elements/statemachined.js",
        b"device": device_target.encode(),
        b"port": str(port).encode(),
    }


@dataclass
class MdnsServiceAdvertisement:
    """One `_statemachined._tcp` registration, for as long as the daemon runs.

    Started with the app and unregistered with it, because a record that
    outlives its daemon is worse than no record: a console shows a rig that
    cannot answer, and the person believes the rig is broken rather than absent.
    """

    port: int
    device_target: str = ""
    version: str = "0.0.0"
    instance_name: str = ""
    #: Where a failure goes. Print by default, so that a daemon started by hand
    #: says why it is not discoverable instead of being quietly invisible.
    report: object = print

    is_advertising: bool = field(default=False, init=False)
    last_failure: str = field(default="", init=False)
    _zeroconf: object | None = field(default=None, init=False)
    _service_info: object | None = field(default=None, init=False)

    def service_name(self) -> str:
        """`<instance>._statemachined._tcp.local.`

        The hostname by default, because that is what somebody standing next to
        the rig would call it. Collisions are Zeroconf's problem and the `id=`
        record is what survives its answer.
        """
        instance = self.instance_name or socket.gethostname().split(".")[0]
        return f"{instance}.{SERVICE_TYPE}"

    def build_service_info(self):
        """The record, as zeroconf wants it. Separate so a test can read it."""
        from zeroconf import ServiceInfo

        rig_identifier = stable_rig_identifier()
        return ServiceInfo(
            SERVICE_TYPE,
            self.service_name(),
            port=self.port,
            properties=text_records_for(
                rig_identifier=rig_identifier,
                port=self.port,
                device_target=self.device_target,
                version=self.version,
            ),
            server=f"{socket.gethostname().split('.')[0]}.local.",
        )

    def start(self) -> bool:
        """Register. Returns whether it worked; never raises."""
        try:
            from zeroconf import Zeroconf

            self._service_info = self.build_service_info()
            self._zeroconf = Zeroconf()
            self._zeroconf.register_service(self._service_info)
        except Exception as exc:  # noqa: BLE001 -- see the module docstring
            self.last_failure = f"{type(exc).__name__}: {exc}"
            self._report(
                f"mDNS: not advertising ({self.last_failure}). "
                "The API is still served; a console will need the address by hand."
            )
            self._close_quietly()
            return False
        self.is_advertising = True
        return True

    def stop(self) -> None:
        """Withdraw the record, then close. Also never raises."""
        if self._zeroconf is not None and self._service_info is not None:
            try:
                self._zeroconf.unregister_service(self._service_info)
            except Exception as exc:  # noqa: BLE001
                self.last_failure = f"{type(exc).__name__}: {exc}"
        self._close_quietly()
        self.is_advertising = False

    def _close_quietly(self) -> None:
        if self._zeroconf is not None:
            try:
                self._zeroconf.close()
            except Exception:  # noqa: BLE001
                pass
        self._zeroconf = None
        self._service_info = None

    def _report(self, text: str) -> None:
        if callable(self.report):
            self.report(text)

File: daemon/src/mdns_service_advertisement.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
//! Telling the network this rig exists.
//!
//! `_statemachined._tcp`, beside vstimd's `_vstimd._tcp` and triald's, so one
//! browse of the local domain finds every braemons daemon on a rig. A Pi whose
//! hostname is generated at boot cannot be hand-configured into a console's URL
//! list, and two rigs called `raspberrypi.local` is the normal case.
//!
//! **The identity is not the hostname.** `/etc/machine-id` survives a rename
//! and is different on every box, so a console that has seen this rig before
//! recognises it afterwards. It is hashed rather than published — a machine-id
//! is meant to be confidential — and hashed exactly as the Python daemon hashes
//! it, so a rig keeps its identity across the cutover.
//!
//! **Advertising is never fatal.** A daemon whose network is down still owns a
//! device and still serves an API somebody may reach by IP, so every failure
//! here is reported and swallowed, and `is_advertising` says which happened.

use std::collections::HashMap;
use std::path::Path;

use sha2::{Digest, Sha256};

pub const SERVICE_TYPE: &str = "_statemachined._tcp.local.";
pub const MACHINE_ID_PATH: &str = "/etc/machine-id";

/// Sixteen hex digits that mean "this box", across renames and reboots.
///
/// Falls back to the hostname where there is no machine-id — a container, a
/// developer's checkout. Weaker, and better than a value that changes every
/// restart, which would make a console show one rig as many.
pub fn stable_rig_identifier(machine_id_path: &Path) -> String {
    let seed = std::fs::read_to_string(machine_id_path)
        .map(|text| text.trim().to_string())
        .unwrap_or_default();
    let seed = if seed.is_empty() { hostname() } else { seed };
    let digest = Sha256::digest(format!("statemachined:{seed}").as_bytes());
    digest.iter().map(|byte| format!("{byte:02x}")).collect::<String>()[..16].to_string()
}

/// The box rather than this daemon on it: the same machine-id, salted
/// `braemons:`, and so the same sixteen digits in every braemons daemon's
/// record on one rig. `id` cannot do this — it is salted with the daemon's own
/// name by design — and a console grouping on it shows each rig once per
/// daemon (`console/docs/PLAN.md` §4).
pub fn rig_identifier(machine_id_path: &Path) -> String {
    let seed = std::fs::read_to_string(machine_id_path)
        .map(|text| text.trim().to_string())
        .unwrap_or_default();
    let seed = if seed.is_empty() { hostname() } else { seed };
    let digest = Sha256::digest(format!("braemons:{seed}").as_bytes());
    digest.iter().map(|byte| format!("{byte:02x}")).collect::<String>()[..16].to_string()
}

/// What a console can act on without opening a connection first.
///
/// Deliberately small: a TXT record is not an API. `elements` is here because
/// the panels' URL is one a console has to construct, and constructing it from
/// a convention rather than from the record is how a console breaks behind a
/// proxy. `api` is carried as the Python daemon carries it, though nothing
/// reads it (`dev/RUST_PORT_STATUS.md`, open questions).
pub fn text_records_for(
    rig_identifier: &str,
    port: u16,
    device_target: &str,
    version: &str,
) -> HashMap<String, String> {
    [
        ("id", rig_identifier.to_string()),
        ("version", version.to_string()),
        ("api", "/api".to_string()),
        ("elements", "/elements/statemachined.js".to_string()),
        ("device", device_target.to_string()),
        ("port", port.to_string()),
    ]
    .into_iter()
    .map(|(key, value)| (key.to_string(), value))
    .collect()
}

/// This box's name, as somebody standing next to the rig would say it.
pub fn hostname() -> String {
    let mut buffer = [0u8; 256];
    // SAFETY: a buffer and its length, and gethostname writes within it.
    let written = unsafe { libc::gethostname(buffer.as_mut_ptr().cast(), buffer.len()) };
    if written != 0 {
        return "statemachined".into();
    }
    let end = buffer.iter().position(|byte| *byte == 0).unwrap_or(buffer.len());
    String::from_utf8_lossy(&buffer[..end]).to_string()
}

/// One `_statemachined._tcp` registration, for as long as the daemon runs.
///
/// Withdrawn with it, because a record that outlives its daemon is worse than
/// none: a console shows a rig that cannot answer.
pub struct MdnsServiceAdvertisement {
    pub port: u16,
    pub device_target: String,
    pub version: String,
    /// The hostname's first label when empty.
    pub instance_name: String,
    pub is_advertising: bool,
    pub last_failure: String,
    daemon: Option<mdns_sd::ServiceDaemon>,
    fullname: Option<String>,
}

impl MdnsServiceAdvertisement {
    pub fn new(port: u16, device_target: &str, version: &str) -> Self {
        Self {
            port,
            device_target: device_target.to_string(),
            version: version.to_string(),
            instance_name: String::new(),
            is_advertising: false,
            last_failure: String::new(),
            daemon: None,
            fullname: None,
        }
    }

    fn short_hostname() -> String {
        hostname().split('.').next().unwrap_or_default().to_string()
    }

    /// `<instance>._statemachined._tcp.local.` — the hostname by default.
    /// Collisions are the responder's problem, and the `id` record is what
    /// survives its answer.
    pub fn service_name(&self) -> String {
        let instance = if self.instance_name.is_empty() {
            Self::short_hostname()
        } else {
            self.instance_name.clone()
        };
        format!("{instance}.{SERVICE_TYPE}")
    }

    /// The record, as the responder wants it. Separate so a test can read it.
    pub fn build_service_info(&self) -> Result<mdns_sd::ServiceInfo, String> {
        let instance = self.service_name();
        let instance = instance.trim_end_matches(&format!(".{SERVICE_TYPE}"));
        let mut records = text_records_for(
            &stable_rig_identifier(Path::new(MACHINE_ID_PATH)),
            self.port,
            &self.device_target,
            &self.version,
        );
        records.insert("rig".into(), rig_identifier(Path::new(MACHINE_ID_PATH)));
        mdns_sd::ServiceInfo::new(
            SERVICE_TYPE,
            instance,
            &format!("{}.local.", Self::short_hostname()),
            "",
            self.port,
            records,
        )
        .map(|info| info.enable_addr_auto())
        .map_err(|problem| problem.to_string())
    }

    /// Register. Returns whether it worked; never fails the daemon.
    pub fn start(&mut self) -> bool {
        let registered = (|| {
            let info = self.build_service_info()?;
            let daemon = mdns_sd::ServiceDaemon::new().map_err(|problem| problem.to_string())?;
            let fullname = info.get_fullname().to_string();
            daemon.register(info).map_err(|problem| problem.to_string())?;
            Ok::<_, String>((daemon, fullname))
        })();
        match registered {
            Ok((daemon, fullname)) => {
                self.daemon = Some(daemon);
                self.fullname = Some(fullname);
                self.is_advertising = true;
                true
            }
            Err(problem) => {
                self.last_failure = problem;
                log::warn!(
                    "mDNS: not advertising ({}). The API is still served; a console will need \
                     the address by hand.",
                    self.last_failure
                );
                false
            }
        }
    }

    /// Withdraw the record, then close. Also never fails.
    pub fn stop(&mut self) {
        if let (Some(daemon), Some(fullname)) = (&self.daemon, &self.fullname) {
            if let Err(problem) = daemon.unregister(fullname) {
                self.last_failure = problem.to_string();
            }
        }
        if let Some(daemon) = self.daemon.take() {
            let _ = daemon.shutdown();
        }
        self.fullname = None;
        self.is_advertising = false;
    }
}

impl Drop for MdnsServiceAdvertisement {
    fn drop(&mut self) {
        self.stop();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_identifier_is_stable_and_is_not_the_machine_id() {
        let directory = tempfile::tempdir().unwrap();
        let machine_id = directory.path().join("machine-id");
        std::fs::write(&machine_id, "dc4f4b06f2d84f6b9e2a7b0c1d2e3f40\n").unwrap();
        let identifier = stable_rig_identifier(&machine_id);
        assert_eq!(identifier, stable_rig_identifier(&machine_id));
        assert_eq!(identifier.len(), 16);
        assert!(!identifier.contains("dc4f4b06"));
    }

    #[test]
    fn the_identifier_is_the_one_the_python_daemon_gave_this_rig() {
        // Computed by `stable_rig_identifier` in Python for this machine-id. A
        // console recognises a rig by it; changing daemons must not rename it.
        let directory = tempfile::tempdir().unwrap();
        let machine_id = directory.path().join("machine-id");
        std::fs::write(&machine_id, "dc4f4b06f2d84f6b9e2a7b0c1d2e3f40\n").unwrap();
        assert_eq!(stable_rig_identifier(&machine_id), "0b60208b6da842d7");
    }

    #[test]
    fn the_rig_identifier_is_the_family_one() {
        // mousewheeld's test asserts the same salting; this is the value both
        // daemons advertise as `rig=` for this machine-id.
        let directory = tempfile::tempdir().unwrap();
        let machine_id = directory.path().join("machine-id");
        std::fs::write(&machine_id, "dc4f4b06f2d84f6b9e2a7b0c1d2e3f40\n").unwrap();
        let rig = rig_identifier(&machine_id);
        assert_eq!(rig.len(), 16);
        assert_ne!(rig, stable_rig_identifier(&machine_id));
    }

    #[test]
    fn two_boxes_get_different_identifiers() {
        let directory = tempfile::tempdir().unwrap();
        std::fs::write(directory.path().join("one"), "1".repeat(32)).unwrap();
        std::fs::write(directory.path().join("two"), "2".repeat(32)).unwrap();
        assert_ne!(
            stable_rig_identifier(&directory.path().join("one")),
            stable_rig_identifier(&directory.path().join("two"))
        );
    }

    #[test]
    fn a_box_with_no_machine_id_still_gets_one() {
        let directory = tempfile::tempdir().unwrap();
        assert_eq!(stable_rig_identifier(&directory.path().join("absent")).len(), 16);
    }

    #[test]
    fn the_text_records_say_where_the_api_and_the_elements_are() {
        let records = text_records_for("abc123", 8081, "/dev/ttyACM0", "1.2.3");
        assert_eq!(records["id"], "abc123");
        assert_eq!(records["api"], "/api");
        assert_eq!(records["elements"], "/elements/statemachined.js");
        assert_eq!(records["device"], "/dev/ttyACM0");
        assert_eq!(records["version"], "1.2.3");
    }

    #[test]
    fn the_service_name_is_under_the_braemons_service_type() {
        let mut advertisement = MdnsServiceAdvertisement::new(8081, "", "0");
        advertisement.instance_name = "rig-3".into();
        assert_eq!(advertisement.service_name(), format!("rig-3.{SERVICE_TYPE}"));
    }

    #[test]
    fn the_record_carries_the_port_it_was_told_to_advertise() {
        let mut advertisement = MdnsServiceAdvertisement::new(9099, "", "0");
        advertisement.instance_name = "rig-3".into();
        let info = advertisement.build_service_info().unwrap();
        assert_eq!(info.get_port(), 9099);
        assert_eq!(info.get_property_val_str("port"), Some("9099"));
    }

    #[test]
    fn stopping_something_that_never_started_is_not_a_failure() {
        let mut advertisement = MdnsServiceAdvertisement::new(8081, "", "0");
        advertisement.stop();
        assert!(!advertisement.is_advertising);
        assert!(advertisement.last_failure.is_empty());
    }
}
